src/taxes/components/TaxCountryDialog/TaxCountryDialog.tsx

```diff
@@ -0,0 +1,119 @@
+import {
+  Dialog,
+  DialogActions,
+  DialogContent,
+  InputAdornment,
+  TextField
+} from "@material-ui/core";
+import VerticalSpacer from "@saleor/apps/components/VerticalSpacer";
+import { CountryFragment } from "@saleor/graphql";
+import { ChangeEvent } from "@saleor/hooks/useForm";
+import { useLocalSearch } from "@saleor/hooks/useLocalSearch";
+import { buttonMessages } from "@saleor/intl";
+import { Button, DialogHeader, SearchIcon } from "@saleor/macaw-ui";
+import { taxesMessages } from "@saleor/taxes/messages";
+import React from "react";
+import { FormattedMessage, useIntl } from "react-intl";
+
+import { useStyles } from "./styles";
+import { TaxCountryDialogLine } from "./TaxCountryDialogLine";
+
+interface TaxCountryDialogProps {
+  open: boolean;
+  countries: CountryFragment[];
+  onConfirm: (countries: CountryFragment[]) => void;
+  onClose: () => void;
+}
+
+export interface CountryFragmentWithState extends CountryFragment {
+  checked: boolean;
+}
+
+export const TaxCountryDialog: React.FC<TaxCountryDialogProps> = ({
+  open,
+  countries,
+  onConfirm,
+  onClose
+}) => {
+  const classes = useStyles();
+  const intl = useIntl();
+
+  const [countriesWithState, setCountriesWithState] = React.useState<
+    CountryFragmentWithState[]
+  >([]);
+  React.useEffect(
+    () =>
+      setCountriesWithState(
+        countries.map(country => ({ checked: false, ...country }))
+      ),
+    [countries]
+  );
```

Comment:
I think we can just use `useStateFromProps` and map it earlier in the parent

Reply:
We can't actually. `useStateFromProps` uses `isEqual` which disables state update on prop dereference, which is required here for proper checkbox state resetting when modal is closed. `open` should be added to useEffect deps and initial value can indeed be mapped earlier in the parent.